searx/engines/google_videos.py

```diff
@@ -1,97 +1,219 @@
-"""
- Google (Videos)
+# SPDX-License-Identifier: AGPL-3.0-or-later
+"""Google (Viedo)
+
+:website:     https://video.google.com
+:provide-api: yes (https://developers.google.com/custom-search/)
+:using-api:   not the offical, since it needs registration to another service
+:results:     HTML
+:stable:      no
+:template:    video.html
+:parse:       url, title, content, thumbnail
+
+For detailed description of the *REST-full* API see: `Query Parameter
+Definitions`_.  Not all parameters can be appied.
+
+.. _admonition:: Content-Security-Policy (CSP)
+
+   This engine needs to allow images from the `data URLs`_ (prefixed with the
+   ``data:` scheme).::
+
+     Header set Content-Security-Policy "img-src 'self' data: ;"
 
- @website     https://www.google.com
- @provide-api yes (https://developers.google.com/custom-search/)
+.. _Query Parameter Definitions:
+   https://developers.google.com/custom-search/docs/xml_results#WebSearch_Query_Parameter_Definitions
+.. _data URLs:
+   https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs
 
- @using-api   no
- @results     HTML
- @stable      no
- @parse       url, title, content, thumbnail
 """
 
-from datetime import date, timedelta
-from json import loads
+# pylint: disable=invalid-name, missing-function-docstring
+
+import re
 from lxml import html
+from flask_babel import gettext
+from searx import logger
+from searx.url_utils import urlencode, urlparse
+from searx.utils import eval_xpath
 from searx.engines.xpath import extract_text
-from searx.url_utils import urlencode
-import re
+
+# pylint: disable=unused-import
+from searx.engines.google import (
+    supported_languages_url
+    ,  _fetch_supported_languages
+)
+# pylint: enable=unused-import
+
+from searx.engines.google import (
+    get_lang_country
+    , google_domains
+    , time_range_dict
+    , filter_mapping
+    , results_xpath
+    , g_section_with_header
+    , title_xpath
+    , href_xpath
+    , content_xpath
+    , suggestion_xpath
+    , spelling_suggestion_xpath
+)
+
+logger = logger.getChild('google video')
 
 # engine dependent config
+
 categories = ['videos']
-paging = True
-safesearch = True
+paging = False
+language_support = True
+use_locale_domain = True
 time_range_support = True
-number_of_results = 10
+safesearch = True
 
-search_url = 'https://www.google.com/search'\
-    '?q={query}'\
-    '&tbm=vid'\
-    '&{search_options}'
-time_range_attr = "qdr:{range}"
-time_range_custom_attr = "cdr:1,cd_min:{start},cd_max{end}"
-time_range_dict = {'day': 'd',
-                   'week': 'w',
-                   'month': 'm'}
+RE_CACHE = {}
 
+def _re(regexpr):
```

Comment:
Because globals need to be evaluated when the engine starts / when the module is imported. Python has many of such constructs.  Evaluate code while module is imported slows down the creation of the runtime environment when a new (interpreter) processes is initialized.

A whole series of reg-expr are required here, so I choose a late evaluating concept combined with a global cache.  BTW it is more readable; to have the regular expression defined where it is used.

Reply:
Install snakeviz yappi :
```sh
. ./local/py3/bin/activate
pip install snakeviz yappi
```

Then create profile.py with the following content (at the root folder, next to README.rst)
```python
import yappi
yappi.set_clock_type("cpu")
yappi.start(builtins=True, profile_threads=True)
try:
    def bench_load_engines():
        import searx.engines
        import searx
        searx.engines.initialize_engines(searx.settings['engines'])

    def bench_webapp():
        import searx.webapp
        searx.webapp.run()

    bench_webapp()
finally:
    yappi.stop()
    threads = yappi.get_thread_stats()
    for thread in threads:
        print(thread)
        func_stats = yappi.get_func_stats(ctx_id=thread.id)
        pstats = yappi.convert2pstats(func_stats)
        pstats.dump_stats('searx-%s-%s.prof' % (thread.id, thread.name))
```

then:
```sh
python profile.py
```
(Ctrl-C after 20 seconds)

```
snakeviz searx-0-_MainThread.prof
```

The regex that takes a lot of times are loaded from this file: https://github.com/asciimoo/searx/blob/master/searx/plugins/https_rewrite.py

Reply:
Wow, looks _cewl_ at its first glance.  To come to an end I would handle performance issues in a separate task (coming soon).

I mark this request as resolved, if you think different, please reopen.

Reply:
>  To come to an end I would handle performance issues in a separate task (coming soon).

Yes ! (and great).

> Because globals need to be evaluated when the engine starts / when the module is imported. Python has many of such constructs. Evaluate code while module is imported slows down the creation of the runtime environment when a new (interpreter) processes is initialized.

The benchmark code was about this sentence: I don't think it will improve the start time but without number it is difficult to get a fact.

And if we add a cache it would be better to add the functions in searx/utils.py, same as https://github.com/asciimoo/searx/blob/master/searx/utils.py#L465 (similar to eval_xpath).

Reply:
I have my own abstraction layer for regular expression with a little more comfort (for more readable code) .. e.g. it caches the matches also (known from Perl)  .. I will add this abstraction later, in the utils or as a separated module.

Do we need to anything more here before we can merge this PR .. did I missed any of your questions (this github UI has some pitfalls for me, especially on big patch series ) ..

If you think its resolved please mark this as resolved / I do not know when this github UI thinks a review process is ended and gives me the right to merge (I am not very experienced in such UIs / sorry) .